File: android/lib/models/voice_command.dart
// Copyright 2026 Daniel Pelikan
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// Voice command codes that can be sent to the desktop.
enum CommandCode {
  enter('ENTER'),
  selectAll('SELECT_ALL'),
  copy('COPY'),
  paste('PASTE'),
  cut('CUT'),
  cancel('CANCEL');

  final String code;
  const CommandCode(this.code);

  /// Parse from spoken text.
  /// Returns null if no command is detected.
  static CommandCode? parse(String text) {
    final lower = text.toLowerCase().trim();
    
    // Enter / new line
    if (_matches(lower, ['new line', 'newline', 'enter', 'next line', 'line break'])) {
      return CommandCode.enter;
    }
    
    // Select all
    if (_matches(lower, ['select all', 'select everything', 'highlight all'])) {
      return CommandCode.selectAll;
    }
    
    // Copy
    if (_matches(lower, ['copy that', 'copy this', 'copy it', 'copy selection'])) {
      return CommandCode.copy;
    }
    
    // Paste
    if (_matches(lower, ['paste', 'paste that', 'paste it', 'paste here'])) {
      return CommandCode.paste;
    }
    
    // Cut
    if (_matches(lower, ['cut that', 'cut this', 'cut it', 'cut selection'])) {
      return CommandCode.cut;
    }
    
    // Cancel
    if (_matches(lower, ['cancel', 'clear', 'never mind', 'nevermind', 'discard'])) {
      return CommandCode.cancel;
    }
    
    return null;
  }

  static bool _matches(String text, List<String> patterns) {
    for (final pattern in patterns) {
      if (text.contains(pattern)) {
        return true;
      }
    }
    return false;
  }
}

/// Local commands that affect the app but aren't sent to desktop.
enum LocalCommand {
  stopListening,
  startListening,
}

/// Check for local commands in text.
LocalCommand? parseLocalCommand(String text) {
  final lower = text.toLowerCase().trim();
  
  if (lower.endsWith('stop listening') || lower.endsWith('pause listening')) {
    return LocalCommand.stopListening;
  }
  
  if (lower.endsWith('start listening') || lower.endsWith('resume listening')) {
    return LocalCommand.startListening;
  }
  
  return null;
}

/// Result of processing spoken text for commands.
class ProcessedSpeech {
  /// Text before any detected command (to be sent as TEXT).
  final String? textBefore;
  
  /// Detected command (to be sent as COMMAND).
  final CommandCode? command;
  
  /// Text after command (to be processed in next iteration).
  final String? textAfter;

  ProcessedSpeech({
    this.textBefore,
    this.command,
    this.textAfter,
  });

  /// Process spoken text to extract commands.
  factory ProcessedSpeech.process(String text) {
    final lower = text.toLowerCase();
    
    // Command patterns with their codes
    final patterns = <String, CommandCode>{
      'new line': CommandCode.enter,
      'newline': CommandCode.enter,
      'enter': CommandCode.enter,
      'next line': CommandCode.enter,
      'line break': CommandCode.enter,
      'select all': CommandCode.selectAll,
      'select everything': CommandCode.selectAll,
      'copy that': CommandCode.copy,
      'copy this': CommandCode.copy,
      'copy it': CommandCode.copy,
      'paste': CommandCode.paste,
      'paste that': CommandCode.paste,
      'cut that': CommandCode.cut,
      'cut this': CommandCode.cut,
      'cancel': CommandCode.cancel,
      'clear': CommandCode.cancel,
      'never mind': CommandCode.cancel,
      'nevermind': CommandCode.cancel,
    };

    // Find the first command in the text
    int firstIndex = -1;
    String? firstPattern;
    CommandCode? firstCommand;

    for (final entry in patterns.entries) {
      final index = lower.indexOf(entry.key);
      if (index != -1 && (firstIndex == -1 || index < firstIndex)) {
        firstIndex = index;
        firstPattern = entry.key;
        firstCommand = entry.value;
      }
    }

    if (firstIndex == -1) {
      // No command found
      return ProcessedSpeech(textBefore: text);
    }

    final textBefore = text.substring(0, firstIndex).trim();
    final afterCommand = firstIndex + firstPattern!.length;
    final textAfter = afterCommand < text.length 
        ? text.substring(afterCommand).trim() 
        : null;

    return ProcessedSpeech(
      textBefore: textBefore.isEmpty ? null : textBefore,
      command: firstCommand,
      textAfter: textAfter?.isEmpty == true ? null : textAfter,
    );
  }

  bool get hasText => textBefore != null && textBefore!.isNotEmpty;
  bool get hasCommand => command != null;
  bool get hasRemainder => textAfter != null && textAfter!.isNotEmpty;
}

File: android/android/app/src/main/java/com/speech2prompt/domain/model/VoiceCommand.kt
package com.speech2prompt.domain.model

/**
 * UI representation of a voice command for display purposes
 */
data class VoiceCommand(
    val trigger: String,
    val description: String,
    val example: String = ""
)

/**
 * Command codes for voice-triggered actions.
 * These are sent to the Linux host for execution.
 */
enum class CommandCode(val code: String) {
    ENTER("ENTER"),
    SELECT_ALL("SELECT_ALL"),
    COPY("COPY"),
    PASTE("PASTE"),
    CUT("CUT"),
    CANCEL("CANCEL");

    companion object {
        fun fromCode(code: String): CommandCode? {
            return entries.find { it.code.equals(code, ignoreCase = true) }
        }
    }
}

/**
 * Result of processing speech text for commands
 */
data class ProcessedSpeech(
    val textBefore: String?,
    val command: CommandCode?,
    val textAfter: String?
) {
    /**
     * Whether there is any text content (before or after command)
     */
    val hasText: Boolean
        get() = !textBefore.isNullOrBlank() || !textAfter.isNullOrBlank()

    /**
     * Whether a command was detected
     */
    val hasCommand: Boolean
        get() = command != null

    /**
     * Whether there is text after the command
     */
    val hasRemainder: Boolean
        get() = !textAfter.isNullOrBlank()

    /**
     * Combined text (before + after), null if no text
     */
    val combinedText: String?
        get() {
            val before = textBefore?.trim() ?: ""
            val after = textAfter?.trim() ?: ""
            val combined = "$before $after".trim()
            return combined.ifEmpty { null }
        }

    companion object {
        /**
         * Create result with only text (no command)
         */
        fun textOnly(text: String): ProcessedSpeech {
            return ProcessedSpeech(
                textBefore = text.trim().ifEmpty { null },
                command = null,
                textAfter = null
            )
        }

        /**
         * Create result with only command (no text)
         */
        fun commandOnly(command: CommandCode): ProcessedSpeech {
            return ProcessedSpeech(
                textBefore = null,
                command = command,
                textAfter = null
            )
        }

        /**
         * Empty result
         */
        val EMPTY = ProcessedSpeech(null, null, null)
    }
}

/**
 * Parser for detecting voice commands in speech text.
 * Matches Flutter's CommandParser implementation.
 */
object CommandParser {
    /**
     * Command patterns mapped to command codes.
     * Patterns are matched case-insensitively.
     */
    private val patterns: Map<String, CommandCode> = mapOf(
        // Enter/newline variations
        "new line" to CommandCode.ENTER,
        "newline" to CommandCode.ENTER,
        "enter" to CommandCode.ENTER,
        "new paragraph" to CommandCode.ENTER,
        "next line" to CommandCode.ENTER,

        // Selection
        "select all" to CommandCode.SELECT_ALL,
        "select everything" to CommandCode.SELECT_ALL,

        // Copy variations
        "copy that" to CommandCode.COPY,
        "copy this" to CommandCode.COPY,
        "copy" to CommandCode.COPY,
        "copy text" to CommandCode.COPY,

        // Paste variations
        "paste" to CommandCode.PASTE,
        "paste that" to CommandCode.PASTE,
        "paste text" to CommandCode.PASTE,

        // Cut variations
        "cut that" to CommandCode.CUT,
        "cut this" to CommandCode.CUT,
        "cut" to CommandCode.CUT,
        "cut text" to CommandCode.CUT,

        // Cancel variations
        "cancel" to CommandCode.CANCEL,
        "cancel that" to CommandCode.CANCEL,
        "never mind" to CommandCode.CANCEL,
        "nevermind" to CommandCode.CANCEL
    )

    /**
     * Sorted patterns for matching (longest first to avoid partial matches)
     */
    private val sortedPatterns: List<Pair<String, CommandCode>> =
        patterns.entries
            .sortedByDescending { it.key.length }
            .map { it.key to it.value }

    /**
     * Parse text to find a command.
     * Returns the first matching command or null.
     */
    fun parse(text: String): CommandCode? {
        val normalized = text.lowercase().trim()
        
        for ((pattern, command) in sortedPatterns) {
            if (normalized == pattern) {
                return command
            }
        }
        return null
    }

    /**
     * Process text to extract command and surrounding text.
     * Handles cases like "hello new line world" -> text="hello", cmd=ENTER, after="world"
     */
    fun processText(text: String): ProcessedSpeech {
        if (text.isBlank()) {
            return ProcessedSpeech.EMPTY
        }

        val normalized = text.lowercase()

        // Try to find a command pattern in the text
        for ((pattern, command) in sortedPatterns) {
            val index = normalized.indexOf(pattern)
            if (index != -1) {
                // Found a command - extract text before and after
                val before = text.substring(0, index).trim()
                val after = text.substring(index + pattern.length).trim()

                return ProcessedSpeech(
                    textBefore = before.ifEmpty { null },
                    command = command,
                    textAfter = after.ifEmpty { null }
                )
            }
        }

        // No command found - return as plain text
        return ProcessedSpeech.textOnly(text)
    }

    /**
     * Check if text contains any command pattern
     */
    fun containsCommand(text: String): Boolean {
        val normalized = text.lowercase()
        return sortedPatterns.any { (pattern, _) ->
            normalized.contains(pattern)
        }
    }

    /**
     * Get all matching commands in text (for complex utterances)
     */
    fun findAllCommands(text: String): List<Pair<CommandCode, IntRange>> {
        val normalized = text.lowercase()
        val results = mutableListOf<Pair<CommandCode, IntRange>>()
        
        for ((pattern, command) in sortedPatterns) {
            var startIndex = 0
            while (true) {
                val index = normalized.indexOf(pattern, startIndex)
                if (index == -1) break
                
                // Check word boundaries to avoid partial matches
                val beforeOk = index == 0 || !normalized[index - 1].isLetter()
                val afterOk = index + pattern.length >= normalized.length ||
                        !normalized[index + pattern.length].isLetter()
                
                if (beforeOk && afterOk) {
                    results.add(command to IntRange(index, index + pattern.length - 1))
                }
                startIndex = index + 1
            }
        }
        
        return results.sortedBy { it.second.first }
    }
}
